cogeo_mosaic/backends/base.py

```diff
@@ -0,0 +1,66 @@
+"""cogeo_mosaic.backend.base: base Backend class."""
+
+from typing import Dict, List
+
+import abc
+from contextlib import AbstractContextManager
+
+from cogeo_mosaic.model import MosaicJSON
+from cogeo_mosaic.backends.utils import get_hash
+
+
+class BaseBackend(AbstractContextManager):
+    """Base Class for cogeo-mosaic backend storage."""
+
+    mosaic_def: MosaicJSON
+
+    @abc.abstractmethod
+    def __init__(self, *args, **kwargs):
+        """Connect to backend"""
+
+    def __enter__(self):
+        """Support using with Context Managers"""
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Support using with Context Managers"""
+
+    @property
+    def metadata(self) -> Dict:
+        """Retrieve Mosaic metadata
+
+        Returns
+        -------
+        MosaicJSON as dict without `tiles` key.
+        """
+        return self.mosaic_def.dict(exclude={"tiles"}, exclude_none=True)
+
+    @abc.abstractmethod
+    def tile(self, x: int, y: int, z: int) -> List[str]:
+        """Retrieve assets for tile."""
+
+    @abc.abstractmethod
+    def point(self, lng: float, lat: float) -> List[str]:
+        """Retrieve assets for point."""
+
+    @abc.abstractmethod
+    def read(self) -> MosaicJSON:
```

Comment:
@kylebarron @geospatial-jeff Should `read` be a private method?
It's not really needed by the user and will/should only be used by the class itself

Reply:
I agree that it would be uncommon for a user to use it, but I don't necessarily think it should be private.

Reply:
so in fact I learn something, `_` is not really **private** but more like semiprivate, while `__` is **private** 

> one _ simply to indicate to other programmers that the attribute or method is intended to be private

ref: https://stackoverflow.com/a/1301369

I do think `read` is should be at least `semiprivate`

Reply:
I don't follow your conclusion... Both `_` and `__` are private, but `__` has special meanings within a class.

I also don't agree that it will never be called by the user. What if someone wanted to implement a custom merging between an existing mosaic and one on a backend? They'd do:
```py
with MosaicBackend(local_mosaic, url=...) as mosaic:
	remote_mosaic = mosaic.read()
```
Because if you call `mosaic.mosaic_def`, it gives you the `local_mosaic` you passed in.

Reply:
> I don't follow your conclusion... Both _ and __ are private, but __ has special meanings within a class.

well it seems that `_` just serve as an indication, while `__` makes the method fully privat to a class

Reply:
```
with MosaicBackend(local_mosaic, url=...) as mosaic:
	remote_mosaic = mosaic.read()
	# logic to combine mosaics
```

gonna have fun with this one.

I guess this could work better with a `mosaic.merge(another_mosaic)`

```python
with CustomMosaicBackend(local_mosaic) as mosaic:
    with MosaicBackend("https://....") as mosaic_2:
	    mosaic.merge(mosaic_2)
```

Reply:
> I guess this could work better with a `mosaic.merge(another_mosaic)`

But then we're getting more and more specific. How should the mosaics be merged? That's a big question with several different possible solutions.

While I'm not opposed to adding such a `merge` method, I think that the user should be able to call `read` to implement custom merging. And by defining `read` as a public method, we're saying that's behavior that's specifically enabled to the user.

Reply:
The point of the backends is to provide a common base for further work to be done on top of. `read` is a basic crud method, and I think it would be wrong to make it private, so as to tell users not to use it. It's fine to write in the docs that most people won't need to use it, but that doesn't imply it should be private.

Reply:
I'm not saying we should add a merge method... you bring that up ;-). I just showed what could be a cleaner way to do instead of passing 2 mosaics (local_mosaic and url=) in the MosaicBackend ...

Reply:
> I'm not saying we should add a merge method... you bring that up

No, I specifically argued for the opposite... that instead of us creating a `merge` method, we should make public the methods that would enable a user to create a merge method.

I don't see what's wrong with calling it `read` but saying in the docstring "a user normally won't call this directly". I think that's quite common in Python packages.

Reply:
> No, I specifically argued for the opposite... that instead of us creating a merge method, we should make public the methods that would enable a user to create a merge method.

In my example, the `merge` was part of the `CustomMosaicBackend`, I won't implement this ;-)

Reply:
at this point, I still think that if a user do need to call `read` it should be in a really advanced case so it shouldn't matter for the user if it's `read` or `_read`. But from simple user perspective, I believe having `read` as a public method could be kinda misleading because in 99.9% of the time people won't need the `read` method.